subprojects/base-services/src/main/java/org/gradle/internal/reflect/ReflectionCache.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.internal.reflect;
+
+import java.lang.ref.WeakReference;
+import java.util.WeakHashMap;
+
+public abstract class ReflectionCache<T> {
+    private final Object lock = new Object();
+    private final WeakHashMap<Class<?>, WeakReference<T>> cache = new WeakHashMap<Class<?>, WeakReference<T>>();
```

Comment:
Using weak references for the value will only work when there's a reference from the class that is the key to the value `T`. In the case of the constructor cache, the `T` is `CachedConstructor[]` which is not reachable from the class (the constructor is, but the `CachedConstructor` instances are not).

I think for this to work, we need strong references to `T` but need to ensure that `T` does not maintain strong references to the class.